cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -763,23 +746,25 @@ Status NumPyConverter::ConvertDecimals() {
   RETURN_NOT_OK(builder.Resize(length_));
 
   const auto& decimal_type = static_cast<const DecimalType&>(*type_);
-  PyObject* Decimal_type_object = Decimal.obj();
+  const int64_t length = length_;
 
-  for (int64_t i = 0; i < length_; ++i) {
+  for (int64_t i = 0; i < length; ++i) {
     object = objects[i];
 
-    if (PyObject_IsInstance(object, Decimal_type_object)) {
-      Decimal128 value;
-      RETURN_NOT_OK(internal::DecimalFromPythonDecimal(object, decimal_type, &value));
-      RETURN_NOT_OK(builder.Append(value));
-    } else if (PandasObjectIsNull(object)) {
-      RETURN_NOT_OK(builder.AppendNull());
-    } else {
+    if (ARROW_PREDICT_FALSE(!internal::PyDecimal_Check(object))) {
       std::stringstream ss;
       ss << "Error converting from Python objects to Decimal: ";
       RETURN_NOT_OK(InvalidConversion(object, "decimal.Decimal", &ss));
       return Status::Invalid(ss.str());
     }
+
+    if (PandasObjectIsNull(object) || internal::PyDecimal_ISNAN(object)) {
```

Comment:
Can `PandasObjectIsNull` return true on a Decimal instance?

Reply:
No, I'll fix that.